velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <deque>
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <TypeKind kind>
+void applyTypedFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+
+    auto search = searchDecoded.valueAt<T>(row);
+
+    for (auto i = 0; i < size; i++) {
+      if (!elementsDecoded.isNullAt(offset + i) &&
+          elementsDecoded.valueAt<T>(offset + i) == search) {
+        flatResult.set(row, i + 1);
+        return;
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+void applyComplexTypeFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  auto elementsBase = elementsDecoded.base();
+
+  auto searchBase = searchDecoded.base();
+  auto searchIndices = searchDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+    auto searchIndex = searchIndices[row];
+
+    for (auto i = 0; i < size; i++) {
+      if (!elementsBase->isNullAt(offset + i) &&
+          elementsBase->equalValueAt(searchBase, offset + i, searchIndex)) {
+        flatResult.set(row, i + 1);
+        return;
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+template <>
+void applyTypedFirstMatch<TypeKind::ARRAY>(
```

Comment:
nit: I wonder if we could use std::enable_if with TypeTraits<kind>::isPrimitiveType to specialize this template for all complex types at once

Reply:
Added a template parameter of std::enable_if and removed the separate specializations for each complex type. Thank you for this suggestion!